### DISCUSSION TITLE
Tricks to increase development speed

### DISCUSSION THREAD

**Discussion by jamesmbaazam** [epiforecasts/EpiNow2#479] [General]
Some usual development and rinse-and-repeat tasks like generating documentation, reloading the package to test new functionality by running `devtools::load_all()`, etc can take time when contributing to EpiNow2. 

Could you share some tips on how you speed up your developing workflow? I'm hoping to put these together under a section of the contribution guide to help myself and future contributors. There have been talks of rebasing on the `cmdstanr` branch for example, to compile the models quickly. What are some other shortcuts/tricks you use?
@sbfnk and @seabbs

**Comment by sbfnk** [comment]
I don't think I have much to offer I'm afraid - when doing lots of work on the stan code I'd indeed sometimes work on branches based on the `cmdstanr` branch (which won't always recompile everything) and rebase to `main` only when done. But for work on R code no recompilation should be necessary. I'm not sure about `devtools::load_all()` but I'm pretty sure `R CMD install` only recompiles if/when the timestamp of the stan model files gets updated.

**Reply by seabbs** [reply]
Yup `devtools::load_all()` caches as well. Unfortunately though for me it produces spuriously slow stan code so I use `R CMD install` instead.

**Comment by seabbs** [comment]
Echoing @sbfnk the answer is sadly no which is a real pain. 

In theory it should be possible for `devtools::document()` to skip compilation (rather than just caching as it currently does) which would help for our edge case but that would need to be implemented upstream of us.

**Comment by jamesmbaazam** [comment]
I've started using the [job](https://lindeloev.github.io/job/) package for running smaller things like tests in the background so that I can work on other stuff.

For example, after fixing something in `estimate_secondary.R`, I would run its test in the background with
```{r}
job::job({devtools::test_active_file(file = "./R/estimate_secondary.R")})
```
This will free up the console for me to continue working on other stuff.

**Reply by seabbs** [reply]
I thought that Rstudio already ran tests and package builds in the background when triggered using the short cuts? This sounds like a good solution though.

**Reply by jamesmbaazam** [reply]
It's always run in my console. Maybe it's a setting somewhere in the IDE? I'll check.

**Reply by seabbs** [reply]
hmm weird. I could be wrong though as I haven't used Rstudio in a while.

**Reply by seabbs** [reply]
![Screenshot 2023-10-23 at 17 12 18](https://github.com/epiforecasts/EpiNow2/assets/20317649/86a166e9-4eff-4938-8dd0-6b5f87616915)
Using command + shift + t I get this for `epinowcast` (my console/terminal are on the right here).

**Comment by jamesmbaazam** [comment]
You're right about using the shortcut. I didn't know that. Thanks.
